src/services/storage/drivers/inMemory/index.ts

```diff
@@ -64,8 +65,11 @@ export class InMemoryStorage implements StorageDriver {
 		}
 
 		public async save(requestData: UrlRequestData): Promise<StoredUrl> {
+			if (requestData.id && this.storage.data.urls.has(requestData.id)) {
+				throw new GeneralError('The specific id you chose is already in use')
```

Comment:
Can you also add that to the Relational driver save method?

If save fails, check if the error is an id conflict and throw this error instead. If not, just throw an empty GeneralError()

Reply:
will do

Reply:
I'm relying on Postgres code for this verification:
https://www.postgresql.org/docs/9.6/errcodes-appendix.html

need to take into count that in the future, if a Relational database will be implemented with services other than Postgres, the error code might be different

- maybe we should check if the id exists before trying to save, and not trying to save and only then check the error code? this approach will suit every relational DB.
@SnirShechter

Reply:
Interesting. Can you check if there's a unified error code or something? I prefer not to make yet another DB query.

If there's no other way, do it. It's not a big performance hit, given that the id check query is only executed for custom ids.

Reply:
Couldn't find a general error code
The way I see it if we want to make sure to throw the same error message, then we have to make another query
If we don't mind the error message content, but we just want to reflect to the user what went wrong, we can leave it as it was... meaning, using the DB constraints without catching and re-throwing an exception
After all, there are a lot of constraints, going into the direction of formalizing each constraint to throw a custom error message will force us to handle (gradually) all the constraints that were implemented by the DB being used.
I think we should avoid it for the time being

What do you think?

Reply:
@lesagi 
We should never throw an error directly from the database. First of all, since it depends on the implementation (i.e Redis vs Postgres etc.), then our client won't be able to get unified errors. Second of all, passing an error from the database might expose sensitive information.

If you want to perform another query (only in the event of custom ids), it should be fine. If you want to ignore it and just throw a general error, that's fine too. Passing the DB error is not a good option.